### ISSUE TITLE
`let foo = null` isn't an error with noImplicitAny

### ISSUE THREAD

**Issue by danvk** [Microsoft/TypeScript#17299]
Maybe I'm misunderstanding something, but I was really surprised by this!

**TypeScript Version:**  2.4.1

**Code**

```ts
let foo = null;
```

with `noImplicitAny`: http://www.typescriptlang.org/play/#src=let%20foo%20=%20null;

**Expected behavior:**

`foo` has an implicit `any` type but isn't flagged with `noImplicitAny`:

<img width="878" alt="screen shot 2017-07-19 at 1 03 45 pm" src="https://user-images.githubusercontent.com/98301/28379743-bfa2e4ea-6c82-11e7-8776-7b3239957962.png">

**Actual behavior:**

Not an error.

**Comment by RyanCavanaugh** [comment]
The implicit any is now tracked in a more complex way that makes the error appear later, e.g.
```
// Error, return type of 'fn' is implicitly any
function fn() {
  let foo = null;
  return foo;
}
```

If you use `foo` in a way that exposes its `any`ness, you should see an error (post a repro if that happens)

**Comment by danvk** [comment]
Here's the broader context:

```ts
import * as React from 'react';

export class Foo extends React.Component<{}, {}> {
  render() {
    let foo = null;

    if (Math.random() > 0.5) {
      foo = <div />;
    }

    return <div>{foo}</div>;
  }
}
```

When I mouse over either of the first two mentions of `foo`, its type is `any`. But when I mouse over the third, it's `JSX.Element | null`, which is correct. So I guess this is WAI, it's just a bit surprising to see what looks like an implicit `any` not trigger a noImplicitAny error!
